feat: new style for payment option in app/webshop

Oleninkaa opened the pull request:
Closes https://github.com/AtB-AS/kundevendt/issues/18186

### If user is **logged in**:


Profile saved cards           |  Preselected last saved payment option
:-------------------------:|:-------------------------:
![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 08 15](https://github.com/user-attachments/assets/b0d25d6c-25be-4590-8846-c21ab1c4465a)|  ![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 08 55](https://github.com/user-attachments/assets/7d471255-ac63-4ea9-9ebd-6d97360ca1d5)

User clicks on `Edit`         |  If selects `New Card`
:-------------------------:|:-------------------------:
![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 09 05](https://github.com/user-attachments/assets/10530df1-94be-4067-9546-9682e9d5c9fb)| ![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 09 21](https://github.com/user-attachments/assets/6e5682c3-c7f0-40e2-9b93-22ff5a9c1354)


### If user **is anonymous**:

Two options to select    |  If selects `New Card`
:-------------------------:|:-------------------------:
![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 10 14](https://github.com/user-attachments/assets/7ef864dd-0f42-4a36-9cc2-8081d2896a5e) |  ![Simulator Screenshot - iPhone 16 Pro Max - 2025-03-05 at 14 10 19](https://github.com/user-attachments/assets/70fb2b07-36a3-43f4-88d1-74a160a31af1)


I have tried to create a `PaymentSelectionCard` component for both **delete** and **edit**  card functionalities, but I have encountered some issues with types. For **delete**, we should use `RecurringPayment`, but for **edit**, we should use `PaymentMethod`.

The main goal was to split one payment selection component into three: `SinglePaymentMethod` for 'Vipps', `MultiplePaymentMethodsRadioSection` for the group of payment methods, and `PaymentSelectionCard` for actions such as delete and edit.

rosvik commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -128,6 +173,107 @@ export const SelectPaymentMethodSheet: React.FC<Props> = ({
+  shouldSave: boolean;
+  onSetShouldSave: (val: boolean) => void;
+  paymentGroup: PaymentMethod[];
+  index: number;
```
Could we replace the `index` param with `testID`?

rosvik reviewed (commented):
Nice! Noticed a few things reading through this, but it looks and works great when testing locally.

Also, there's a bit of commented out code here. Better to remove that, the we can re-implement it later if needed.

jorunnl commented on src/components/bottom-sheet/BottomSheetContainer.tsx:
```diff
@@ -31,7 +32,7 @@ export function BottomSheetContainer({
         focusTitleOnLoad={focusTitleOnLoad}
       />
-      {children}
+      <ScrollView>{children}</ScrollView>
```
I think this should not be done for the entire BottomSheetContainer component. See if it can only be done where you need it.

jorunnl reviewed (commented):
Looks and works very well 👍 Some feedback in comments. Let me know if we should look at them together 🤗

rosvik commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -128,6 +165,107 @@ export const SelectPaymentMethodSheet: React.FC<Props> = ({
   );
 };
 
+type MultiplePaymentMethodsProps = {
+  paymentMethod?: PaymentMethod;
+  selected: boolean;
+  onSelect: (value: PaymentMethod[]) => void;
+  shouldSave: boolean;
+  onSetShouldSave: (val: boolean) => void;
+  paymentGroup: PaymentMethod[];
+  testID?: string;
+};
+
+const MultiplePaymentMethodsRadioSection: React.FC<
```
We had a talk about this a while back, and decided to try splitting them up, to avoid having too much complexity in one component. There's some differences between them in functionality, even though they look similar.

rosvik commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -52,59 +56,92 @@ export const SelectPaymentMethodSheet: React.FC<Props> = ({
+            }
+            onSelect={() => {
+              setShouldSave(true);
+              setSelectedMethod(multiplePaymentMethod[0]);
```
I'm a bit unsure if saving the first payment method here is the best way to do it. Could we use a different state to more clearly communicate what is selected? Something like this maybe:

```ts
type PaymentSelection = {
	kind: 'new_recurring' | 'saved_recurring' | 'non_recurring',
	paymentMethod: PaymentMethod | undefined,
}
```

I haven't looked into the details here, so you might have to change it a bit, but do you think something like it could work? If it instead makes it harder to work with, feel free to ignore 😄

Oleninkaa commented:
I forgot to mention that there are also some changes on this screen. @rosvik @jorunnl 

![simulator_screenshot_3E8CE45F-0D36-4126-B0A3-C69E0BC1A826](https://github.com/user-attachments/assets/f7a60bfb-fb16-464b-b5c5-1f42e49d6676)

reidzeibel commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -418,6 +403,30 @@ const SinglePaymentMethod: React.FC<PaymentMethodProps> = ({
   );
 };
 
+function defineSelection(
```
`defineSelection` seems to be too generic, I would pick something more specific, for example `getPaymentSelection`

Oleninkaa commented:
When a user pays with a new card and selects to save the card, the next time they visit the `Root_PurchaseConfirmationScreen`, they will see the previously saved payment method. @rosvik Johannes suggested this simple solution. We just added a **delay** to make it work for a newly added card.

I also removed unnecessary elements for the single payment method that I hadn't noticed before. The `PaymentType.PaymentCard` is used when a user selects `MultiplePaymentMethods` and adds a new card in NETS. After the payment flow, we determine the actual type of the card. This is because NETS doesn't really care what card is typed in.

reidzeibel commented on src/stacks-hierarchy/saved-payment-utils.ts:
```diff
@@ -131,6 +131,7 @@ export const saveLastUsedRecurringPaymentOrType = async (
     });
   } else {
     try {
+      await new Promise((resolve) => setTimeout(resolve, 3000));
```
@Oleninkaa why do we need this timeout? 🤔

reidzeibel reviewed (approved):
Looks good, but need to wait for code owners for merging!

marius-at-atb commented on src/stacks-hierarchy/utils.ts:
```diff
@@ -15,7 +15,7 @@ export function getExpireDate(iso: string): string {
   // Example: The card expires the moment the date is 02.2021, but the date on the card is 01.2021
   date.setDate(date.getDate() - 1);
-  return format(date, 'MM/yy');
+  return format(date, 'dd.MM.yy');
```
I can't find where `getExpireDate` is used. Or is it just me?
We can delete it if unused. Otherwise we should probably add some tests for it, to confirm that it works as intended.

Oleninkaa replied:
We will use it for the saved card/NETS expiry messages when we implement this feature

marius-at-atb reviewed (approved):
Great work!
All the comments are syntactic sugar/best practice suggestions/clean-ups, not blocking :)

Oleninkaa commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -92,62 +127,170 @@ export const SelectPaymentMethodSheet: React.FC<Props> = ({
   shouldSave,
   onSetShouldSave,
-  index,
+  paymentGroup,
```
I don't think it's possible, since we need this property later in the function `getPaymentSelection(paymentMethod, shouldSave)` 😮

marius-at-atb commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/components/SelectPaymentMethodSheet.tsx:
```diff
@@ -234,8 +229,11 @@ const MultiplePaymentMethodsRadioSection: React.FC<
               <View style={styles.paymentIconsGroup}>
-                {paymentGroup.map((payment) => (
-                  <PaymentBrand paymentType={payment.paymentType} />
+                {paymentMethodsInGroup.map((payment) => (
```
Typically the variable names in map functions should match the array items.
In this case, `paymentMethodInGroup` or just `paymentMethod` would be more precise than just `payment`.

reidzeibel replied:
@Oleninkaa If you fix this, since it's very simple, no need to request for another re-review, you can safely merge :)

marius-at-atb reviewed (approved):
Very nice!
